README

slink2dali, a SeedLink to DataLink client

For usage information see the slink2dali(1) man page in the 'doc'
directory.

-- Building/Installing --

In most environments a simple 'make' will compile the program.

SunOS/Solaris:

In order to compile under Solaris the 'src/Makefile' needs to be edited.
See the file for instructions.

For further installation simply copy the resulting binary and man page
(in the 'doc' directory) to appropriate system directories.

-- Licensing --

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License (GNU-GPL) for more details.  The GNU-GPL and
further information can be found here: http://www.gnu.org/


ChangeLog
2022.013: 0.8RC
	- Add -nt, -nd, -k options to control network timeout, network
	re-connection delay and keep-alive intervals.
	- Update libmseed to 2.19.6
	- Update libslink to 2.6
	- Update libdali to 1.8

2012.310: 0.7
	- Do not forward SL packet types of SLNUM and SLCHA (value 6),
	these internal request packet types can cause problems downstream.

2011.117: 0.6
	- Update libmseed to 2.6.
	- Add -d option to configure SeedLink connection in dial-up mode.

2011.004: 0.5
	- Update libdali to 1.4.

2011.003: 0.4
	- Update libdali to 1.3.

2010.227: 0.3
	- Pass all SeedLink packet types to DataLink server except INFO and KEEP.

2010.171: 0.2
	- Add -N option to change network codes in forwarded SEED records.

2010.117: 0.1
	- Initial version.


libdali/libdali.h
/***********************************************************************//**
 * @file libdali.h
 * 
 * Interface declarations for the DataLink library (libdali).
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public License
 * as published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License (GNU-LGPL) for more details.  The
 * GNU-LGPL and further information can be found here:
 * http://www.gnu.org/
 *
 * @author Chad Trabant, IRIS Data Management Center
 *
 * modified: 2011.003
 ***************************************************************************/

#ifndef LIBDALI_H
#define LIBDALI_H 1

#ifdef __cplusplus
extern "C" {
#endif

#include "portable.h"

#define LIBDALI_VERSION "1.3"        /**< libdali version */
#define LIBDALI_RELEASE "2011.003"   /**< libdali release date */

#define MAXPACKETSIZE       16384    /**< Maximum packet size for libdali */
#define MAXREGEXSIZE        16384    /**< Maximum regex pattern size */
#define MAX_LOG_MSG_LENGTH  200      /**< Maximum length of log messages */

#define LIBDALI_POSITION_EARLIEST -2 /**< Earliest position in the buffer */
#define LIBDALI_POSITION_LATEST   -3 /**< Latest position in the buffer */

/** Maximium stream ID string length */
#define MAXSTREAMID 60

/* Return values for dl_collect() and dl_collect_nb() */
#define DLERROR    -1      /**< Error occurred */
#define DLENDED     0      /**< Connection terminated */
#define DLPACKET    1      /**< Packet returned */
#define DLNOPACKET  2      /**< No packet for non-blocking dl_collect_nb() */

/** High precision time tick interval as 1/modulus seconds *
 * Default modulus of 1000000 defines tick interval as a microsecond */
#define DLTMODULUS 1000000

/** Error code for routines that normally return a high precision time.
 * The time value corresponds to '1902/1/1 00:00:00.000000' with the
 * default DLTMODULUS */
#define DLTERROR -2145916800000000LL

/** Macro to scale a Unix/POSIX epoch time to a high precision time */
#define DL_EPOCH2DLTIME(X) X * (dltime_t) DLTMODULUS
/** Macro to scale a high precision time to a Unix/POSIX epoch time */
#define DL_DLTIME2EPOCH(X) X / DLTMODULUS
  
/** Data type for high-precision time values.
 *  Require a large (>= 64-bit) integer type */
typedef int64_t dltime_t;

/** Logging parameters */
typedef struct DLLog_s
{
  void (*log_print)();          /**< Function pointer for log message printing */
  const char *logprefix;        /**< Log message prefix */
  void (*diag_print)();         /**< Function pointer for diagnostic/error message printing */
  const char *errprefix;        /**< Error message prefix */
  int  verbosity;               /**< Verbosity level */
} DLLog;

/** DataLink connection parameters */
typedef struct DLCP_s
{
  char        addr[100];        /**< The host:port of DataLink server */
  char        clientid[200];    /**< Client program ID as "progname:username:pid:arch", see dlp_genclientid() */
  int         keepalive;        /**< Interval to send keepalive/heartbeat (seconds) */
  int         iotimeout;        /**< Timeout for network I/O operations (seconds) */
  
  /* Connection parameters maintained internally */
  int         link;		/**< The network socket descriptor, maintained internally */
  float       serverproto;      /**< Server version of the DataLink protocol, maintained internally */
  int32_t     maxpktsize;       /**< Maximum packet size for server, maintained internally */
  int8_t      writeperm;        /**< Write permission status from server, maintained internally */
  int64_t     pktid;            /**< Packet ID of last packet received, maintained internally */
  dltime_t    pkttime;          /**< Packet time of last packet received, maintained internally */
  int8_t      keepalive_trig;   /**< Send keepalive trigger, maintained internally */
  dltime_t    keepalive_time;   /**< Keepalive time stamp, maintained internally */
  int8_t      terminate;        /**< Boolean flag to control connection termination, maintained internally */
  int8_t      streaming;        /**< Boolean flag to indicate streaming status, maintained internally */
  
  DLLog      *log;              /**< Logging parameters, maintained internally */
} DLCP;

/** DataLink packet */
typedef struct DLPacket_s
{
  char        streamid[MAXSTREAMID]; /**< Stream ID */
  int64_t     pktid;            /**< Packet ID */
  dltime_t    pkttime;          /**< Packet time */
  dltime_t    datastart;        /**< Data start time */
  dltime_t    dataend;          /**< Data end time */
  int32_t     datasize;         /**< Data size in bytes */
} DLPacket;


/* connection.c */
extern DLCP *  dl_newdlcp (char *address, char *progname);
extern void    dl_freedlcp (DLCP *dlconn);
extern int     dl_exchangeIDs (DLCP *dlconn, int parseresp);
extern int64_t dl_position (DLCP *dlconn, int64_t pktid, dltime_t pkttime);
extern int64_t dl_position_after (DLCP *dlconn, dltime_t datatime);
extern int64_t dl_match (DLCP *dlconn, char *matchpattern);
extern int64_t dl_reject (DLCP *dlconn, char *rejectpattern);
extern int64_t dl_write (DLCP *dlconn, void *packet, int packetlen, char *streamid,
			 dltime_t datastart, dltime_t dataend, int ack);
extern int     dl_read (DLCP *dlconn, int64_t pktid, DLPacket *packet,
			void *packetdata, size_t maxdatasize);
extern int     dl_getinfo (DLCP *dlconn, const char *infotype, char *infomatch,
			   char **infodata, size_t maxinfosize);
extern int     dl_collect (DLCP *dlconn, DLPacket *packet, void *packetdata,
			   size_t maxdatasize, int8_t endflag);
extern int     dl_collect_nb (DLCP *dlconn, DLPacket *packet, void *packetdata,
			      size_t maxdatasize, int8_t endflag);
extern int     dl_handlereply (DLCP *dlconn, void *buffer, int buflen, int64_t *value);
extern void    dl_terminate (DLCP *dlconn);

/* config.c */
extern char   *dl_read_streamlist (DLCP *dlconn, const char *streamfile);

/* network.c */
extern int     dl_connect (DLCP *dlconn);
extern void    dl_disconnect (DLCP *dlconn);
extern int     dl_senddata (DLCP *dlconn, void *buffer, size_t sendlen);
extern int     dl_sendpacket (DLCP *dlconn, void *headerbuf, size_t headerlen,
			      void *databuf, size_t datalen,
			      void *respbuf, int resplen);
extern int     dl_recvdata (DLCP *dlconn, void *buffer, size_t readlen, uint8_t blockflag);
extern int     dl_recvheader (DLCP *dlconn, void *buffer, size_t buflen, uint8_t blockflag);

/* timeutils.c */
extern int     dl_doy2md (int year, int jday, int *month, int *mday);
extern int     dl_md2doy (int year, int month, int mday, int *jday);
extern char   *dl_dltime2isotimestr (dltime_t dltime, char *isotimestr, int8_t subseconds);
extern char   *dl_dltime2mdtimestr (dltime_t dltime, char *mdtimestr, int8_t subseconds);
extern char   *dl_dltime2seedtimestr (dltime_t dltime, char *seedtimestr, int8_t subseconds);
extern dltime_t dl_time2dltime (int year, int day, int hour, int min, int sec, int usec);
extern dltime_t dl_seedtimestr2dltime (char *seedtimestr);
extern dltime_t dl_timestr2dltime (char *timestr);

/* genutils.c */
extern int     dl_splitstreamid (char *streamid, char *w, char *x, char *y, char *z, char *type);
extern int     dl_bigendianhost (void);
extern double  dl_dabs (double value);
extern int     dl_readline (int fd, char *buffer, int buflen);

/* logging.c */
extern int     dl_log (int level, int verb, ...);
extern int     dl_log_r (const DLCP *dlconn, int level, int verb, ...);
extern int     dl_log_rl (DLLog *log, int level, int verb, ...);
extern void    dl_loginit (int verbosity,
			   void (*log_print)(char*), const char *logprefix,
			   void (*diag_print)(char*), const char *errprefix);
extern void    dl_loginit_r (DLCP *dlconn, int verbosity,
			     void (*log_print)(char*), const char *logprefix,
			     void (*diag_print)(char*), const char *errprefix);
extern DLLog  *dl_loginit_rl (DLLog *log, int verbosity,
			      void (*log_print)(char*), const char *logprefix,
			      void (*diag_print)(char*), const char *errprefix);

/* statefile.c */
extern int  dl_recoverstate (DLCP *dlconn, const char *statefile);
extern int  dl_savestate (DLCP *dlconn, const char *statefile);

/* strutils.c */

/* For a linked list of strings, as filled by strparse() */
typedef struct DLstrlist_s {
  char               *element;
  struct DLstrlist_s *next;
} DLstrlist;

extern int  dl_strparse (const char *string, const char *delim, DLstrlist **list);
extern int  dl_strncpclean (char *dest, const char *source, int length);
extern int  dl_addtostring (char **string, char *add, char *delim, int maxlen);


#ifdef __cplusplus
}
#endif
 
#endif /* LIBDALI_H */


libslink/globmatch.c
/*
 * robust glob pattern matcher
 * ozan s. yigit/dec 1994
 * public domain
 *
 * glob patterns:
 *	*	matches zero or more characters
 *	?	matches any single character
 *	[set]	matches any character in the set
 *	[^set]	matches any character NOT in the set
 *		where a set is a group of characters or ranges. a range
 *		is written as two characters seperated with a hyphen: a-z denotes
 *		all characters between a to z inclusive.
 *	[-set]	set matches a literal hypen and any character in the set
 *	[]set]	matches a literal close bracket and any character in the set
 *
 *	char	matches itself except where char is '*' or '?' or '['
 *	\char	matches char, including any pattern character
 *
 * examples:
 *	a*c		ac abc abbc ...
 *	a?c		acc abc aXc ...
 *	a[a-z]c		aac abc acc ...
 *	a[-a-z]c	a-c aac abc ...
 *
 * $Log: not supported by cvs2svn $
 * Revision 1.8  2009/11/14 10:10:56  chad
 * *** empty log message ***
 *
 * Revision 1.1  2007/10/09 00:47:17  chad
 * add extended reply message handling
 *
 * Revision 1.4sl  2007/06/18  12:47:00  ct
 * Rename globmatch to sl_globmatch for integration into libslink.
 *
 * Revision 1.4  2004/12/26  12:38:00  ct
 * Changed function name (amatch -> globmatch), variables and
 * formatting for clarity.  Also add matching header globmatch.h.
 *
 * Revision 1.3  1995/09/14  23:24:23  oz
 * removed boring test/main code.
 *
 * Revision 1.2  94/12/11  10:38:15  oz
 * charset code fixed. it is now robust and interprets all
 * variations of charset [i think] correctly, including [z-a] etc.
 * 
 * Revision 1.1  94/12/08  12:45:23  oz
 * Initial revision
 */


#include "globmatch.h"

#define SL_GLOBMATCH_TRUE    1
#define SL_GLOBMATCH_FALSE   0


/***********************************************************************
 * sl_globmatch:
 *
 * Check if a string matches a globbing pattern.
 *
 * Return 0 if string does not match pattern and non-zero otherwise.
 **********************************************************************/
int
sl_globmatch (char *string, char *pattern)
{
  int negate;
  int match;
  int c;
  
  while ( *pattern )
    {
      if ( !*string && *pattern != '*' )
	return SL_GLOBMATCH_FALSE;
      
      switch ( c = *pattern++ )
	{
	  
	case '*':
	  while ( *pattern == '*' )
	    pattern++;
	  
	  if ( !*pattern )
	    return SL_GLOBMATCH_TRUE;
	  
	  if ( *pattern != '?' && *pattern != '[' && *pattern != '\\' )
	    while ( *string && *pattern != *string )
	      string++;
	  
	  while ( *string )
	    {
	      if ( sl_globmatch(string, pattern) )
		return SL_GLOBMATCH_TRUE;
	      string++;
	    }
	  return SL_GLOBMATCH_FALSE;
	  
	case '?':
	  if ( *string )
	    break;
	  return SL_GLOBMATCH_FALSE;
	  
	  /* set specification is inclusive, that is [a-z] is a, z and
	   * everything in between. this means [z-a] may be interpreted
	   * as a set that contains z, a and nothing in between.
	   */
	case '[':
	  if ( *pattern != SL_GLOBMATCH_NEGATE )
	    negate = SL_GLOBMATCH_FALSE;
	  else
	    {
	      negate = SL_GLOBMATCH_TRUE;
	      pattern++;
	    }
	  
	  match = SL_GLOBMATCH_FALSE;
	  
	  while ( !match && (c = *pattern++) )
	    {
	      if ( !*pattern )
		return SL_GLOBMATCH_FALSE;
	      
	      if ( *pattern == '-' ) 	/* c-c */
		{
		  if ( !*++pattern )
		    return SL_GLOBMATCH_FALSE;
		  if ( *pattern != ']' )
		    {
		      if ( *string == c || *string == *pattern ||
			   ( *string > c && *string < *pattern ) )
			match = SL_GLOBMATCH_TRUE;
		    }
		  else
		    {		/* c-] */
		      if ( *string >= c )
			match = SL_GLOBMATCH_TRUE;
		      break;
		    }
		}
	      else			/* cc or c] */
		{
		  if ( c == *string )
		    match = SL_GLOBMATCH_TRUE;
		  if ( *pattern != ']' )
		    {
		      if ( *pattern == *string )
			match = SL_GLOBMATCH_TRUE;
		    }
		  else
		    break;
		}
	    } 
	  
	  if ( negate == match )
	    return SL_GLOBMATCH_FALSE;
	  
	  /*
	   * if there is a match, skip past the charset and continue on
	   */
	  while ( *pattern && *pattern != ']' )
	    pattern++;
	  if ( !*pattern++ )	/* oops! */
	    return SL_GLOBMATCH_FALSE;
	  break;
	  
	case '\\':
	  if ( *pattern )
	    c = *pattern++;
	default:
	  if ( c != *string )
	    return SL_GLOBMATCH_FALSE;
	  break;
	}
      
      string++;
    }
  
  return !*string;
}


src/slink2dali.c
/***************************************************************************
 * slink2dali.c
 *
 * SeedLink to DataLink.
 *
 * Connect to a SeedLink server and forward data to a DataLink server.
 *
 * Written by Chad Trabant
 *   IRIS Data Management Center
 *
 * modified 2012.310
 ***************************************************************************/

#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include <libdali.h>
#include <libmseed.h>
#include <libslink.h>

#define PACKAGE "slink2dali"
#define VERSION "0.8"

static int sendrecord (char *record, int reclen);
static int parameter_proc (int argcount, char **argvec);
static char *getoptval (int argcount, char **argvec, int argopt);
static void term_handler (int sig);
static void print_timelogc (const char *msg);
static void print_timelog (char *msg);
static void usage (void);

static short int verbose = 0;   /* Flag to control general verbosity */
static int stateint      = 0;   /* Packet interval to save statefile */
static char *netcode     = 0;   /* Change all SEED newtork codes to netcode */
static char *statefile   = 0;   /* State file for saving/restoring stream states */
static int writeack      = 0;   /* Flag to control the request for write acks */
static int dialup        = 0;   /* Flag to control SeedLink dialup mode */
static int keepalive     = 300; /* Interval to send keepalive/heartbeat (secs) */
static int netto         = 600; /* Network timeout (secs) */
static int netdly        = 30;  /* Network reconnect delay (secs) */

static SLCD *slconn;
static DLCP *dlconn;

int
main (int argc, char **argv)
{
  SLpacket *slpack;
  int seqnum;
  int ptype;
  int packetcnt = 0;

  /* The following is dependent on the packet type values in libslink.h */
  char *type[] = {"Data", "Detection", "Calibration", "Timing",
                  "Message", "General", "Request", "Info",
                  "Info (terminated)", "KeepAlive"};

  /* Signal handling, use POSIX calls with standardized semantics */
  struct sigaction sa;

  sa.sa_flags = SA_RESTART;
  sigemptyset (&sa.sa_mask);

  sa.sa_handler = term_handler;
  sigaction (SIGINT, &sa, NULL);
  sigaction (SIGQUIT, &sa, NULL);
  sigaction (SIGTERM, &sa, NULL);

  sa.sa_handler = SIG_IGN;
  sigaction (SIGHUP, &sa, NULL);
  sigaction (SIGPIPE, &sa, NULL);

  /* Process specified parameters */
  if (parameter_proc (argc, argv) < 0)
  {
    fprintf (stderr, "Argument processing failed\n");
    fprintf (stderr, "Try '-h' for detailed help\n");
    return -1;
  }

  /* Connect to DataLink server */
  if (dl_connect (dlconn) < 0)
  {
    sl_log (2, 0, "Error connecting to DataLink server\n");
    return -1;
  }

  /* Loop with the connection manager */
  while (sl_collect (slconn, &slpack))
  {
    ptype  = sl_packettype (slpack);
    seqnum = sl_sequence (slpack);

    if (verbose > 1)
    {
      if (ptype == SLKEEP)
        sl_log (1, 0, "Keep alive packet received\n");
      else
        sl_log (1, 0, "Received %s packet, SeedLink sequence %d\n",
                type[ptype], seqnum);
    }

    /* Send record to the DataLink server if not internal types, INFO or keep alive */
    if (ptype >= SLDATA && ptype < SLNUM)
    {
      while (sendrecord ((char *)&slpack->msrecord, SLRECSIZE))
      {
        if (verbose)
          sl_log (1, 0, "Re-connecting to DataLink server\n");

        /* Re-connect to DataLink server and sleep if error connecting */
        if (dlconn->link != -1)
          dl_disconnect (dlconn);

        if (dl_connect (dlconn) < 0)
        {
          sl_log (2, 0, "Error re-connecting to DataLink server, sleeping 10 seconds\n");
          sleep (10);
        }

        if (slconn->terminate)
          break;
      }

      packetcnt++;
    }

    /* Save intermediate state files */
    if (statefile && stateint)
    {
      if (++packetcnt >= stateint)
      {
        sl_savestate (slconn, statefile);
        packetcnt = 0;
      }
    }
  }

  /* Shut down the connection to SeedLink server */
  if (slconn->link != -1)
    sl_disconnect (slconn);

  /* Shut down the connection to DataLink server */
  if (dlconn->link != -1)
    dl_disconnect (dlconn);

  /* Save state file if specified */
  if (statefile)
    sl_savestate (slconn, statefile);

  return 0;
} /* End of main() */

/***************************************************************************
 * sendrecord:
 *
 * Send the specified record to the DataLink server.
 *
 * Returns 0 on success, and -1 on failure
 ***************************************************************************/
static int
sendrecord (char *record, int reclen)
{
  static MSRecord *msr = NULL;
  hptime_t endtime;
  char streamid[100];
  int rv;

  /* Translate network code if supplied */
  if (netcode)
  {
    ms_strncpopen (((struct fsdh_s *)record)->network, netcode, 2);
  }

  /* Parse Mini-SEED header */
  if ((rv = msr_unpack (record, reclen, &msr, 0, 0)) != MS_NOERROR)
  {
    ms_recsrcname (record, streamid, 0);
    sl_log (2, 0, "Error unpacking %s: %s", streamid, ms_errorstr (rv));
    return -1;
  }

  /* Generate stream ID for this record: NET_STA_LOC_CHAN/MSEED */
  msr_srcname (msr, streamid, 0);
  strcat (streamid, "/MSEED");

  /* Determine high precision end time */
  endtime = msr_endtime (msr);

  /* Send record to server */
  if (dl_write (dlconn, record, reclen, streamid, msr->starttime, endtime, writeack) < 0)
  {
    return -1;
  }

  return 0;
} /* End of sendrecord() */

/***************************************************************************
 * parameter_proc:
 *
 * Process the command line parameters.
 *
 * Returns 0 on success, and -1 on failure.
 ***************************************************************************/
static int
parameter_proc (int argcount, char **argvec)
{
  char *sladdress = 0;
  char *dladdress = 0;
  int error       = 0;

  char *streamfile  = 0; /* stream list file for configuring streams */
  char *multiselect = 0;
  char *selectors   = 0;
  char *timewin     = 0;
  char *tptr;

  SLstrlist *timelist; /* split the time window arg */

  if (argcount <= 1)
    error++;

  /* Process all command line arguments */
  for (optind = 1; optind < argcount; optind++)
  {
    if (strcmp (argvec[optind], "-V") == 0)
    {
      fprintf (stderr, "%s version: %s\n", PACKAGE, VERSION);
      exit (0);
    }
    else if (strcmp (argvec[optind], "-h") == 0)
    {
      usage ();
      exit (0);
    }
    else if (strncmp (argvec[optind], "-v", 2) == 0)
    {
      verbose += strspn (&argvec[optind][1], "v");
    }
    else if (strcmp (argvec[optind], "-l") == 0)
    {
      streamfile = getoptval (argcount, argvec, optind++);
    }
    else if (strcmp (argvec[optind], "-s") == 0)
    {
      selectors = getoptval (argcount, argvec, optind++);
    }
    else if (strcmp (argvec[optind], "-S") == 0)
    {
      multiselect = getoptval (argcount, argvec, optind++);
    }
    else if (strcmp (argvec[optind], "-d") == 0)
    {
      dialup = 1;
    }
    else if (strcmp (argvec[optind], "-N") == 0)
    {
      netcode = getoptval (argcount, argvec, optind++);
    }
    else if (strcmp (argvec[optind], "-x") == 0)
    {
      statefile = getoptval (argcount, argvec, optind++);
    }
    else if (strcmp (argvec[optind], "-tw") == 0)
    {
      timewin = getoptval (argcount, argvec, optind++);
    }
    else if (strcmp (argvec[optind], "-nt") == 0)
    {
      netto = atoi (getoptval (argcount, argvec, optind++));
    }
    else if (strcmp (argvec[optind], "-nd") == 0)
    {
      netdly = atoi (getoptval (argcount, argvec, optind++));
    }
    else if (strcmp (argvec[optind], "-k") == 0)
    {
      keepalive = atoi (getoptval (argcount, argvec, optind++));
    }
    else if (strncmp (argvec[optind], "-", 1) == 0)
    {
      fprintf (stderr, "Unknown option: %s\n", argvec[optind]);
      exit (1);
    }
    else if (!sladdress)
    {
      sladdress = argvec[optind];
    }
    else if (!dladdress)
    {
      dladdress = argvec[optind];
    }
    else
    {
      fprintf (stderr, "Unknown option: %s\n", argvec[optind]);
      exit (1);
    }
  }

  /* Make sure a SeedLink server was specified */
  if (!sladdress)
  {
    fprintf (stderr, "No SeedLink server specified\n\n");
    fprintf (stderr, "%s version %s\n\n", PACKAGE, VERSION);
    fprintf (stderr, "Usage: %s [options] slhost dlhost\n\n", PACKAGE);
    fprintf (stderr, "Try '-h' for detailed help\n");
    exit (1);
  }

  /* Make sure a DataLink server was specified */
  if (!dladdress)
  {
    fprintf (stderr, "No DataLink server specified\n\n");
    fprintf (stderr, "%s version %s\n\n", PACKAGE, VERSION);
    fprintf (stderr, "Usage: %s [options] slhost dlhost\n\n", PACKAGE);
    fprintf (stderr, "Try '-h' for detailed help\n");
    exit (1);
  }

  /* Allocate and initialize SeedLink connection description */
  if (!(slconn = sl_newslcd ()))
  {
    fprintf (stderr, "Cannot allocation SeedLink descriptor\n");
    exit (1);
  }

  slconn->sladdr    = sladdress;
  slconn->netto     = netto;
  slconn->netdly    = netdly;
  slconn->keepalive = keepalive;

  if (dialup)
    slconn->dialup = 1;

  /* Allocate and initialize DataLink connection description */
  if (!(dlconn = dl_newdlcp (dladdress, argvec[0])))
  {
    fprintf (stderr, "Cannot allocation DataLink descriptor\n");
    exit (1);
  }

  /* Initialize the verbosity for the sl_log function */
  sl_loginit (verbose, &print_timelogc, NULL, &print_timelogc, NULL);

  /* Initialize the verbosity for the dl_log function */
  dl_loginit (verbose, &print_timelog, "", &print_timelog, "");

  /* Set stdout (where logs go) to always flush after a newline */
  setvbuf (stdout, NULL, _IOLBF, 0);

  /* Report the program version */
  sl_log (1, 0, "%s version: %s\n", PACKAGE, VERSION);

  /* If errors then report the usage message and quit */
  if (error)
  {
    usage ();
    exit (1);
  }

  /* Load the stream list from a file if specified */
  if (streamfile)
    sl_read_streamlist (slconn, streamfile, selectors);

  /* Split the time window argument */
  if (timewin)
  {

    if (strchr (timewin, ':') == NULL)
    {
      sl_log (2, 0, "time window not in begin:[end] format\n");
      return -1;
    }

    if (sl_strparse (timewin, ":", &timelist) > 2)
    {
      sl_log (2, 0, "time window not in begin:[end] format\n");
      sl_strparse (NULL, NULL, &timelist);
      return -1;
    }

    if (strlen (timelist->element) == 0)
    {
      sl_log (2, 0, "time window must specify a begin time\n");
      sl_strparse (NULL, NULL, &timelist);
      return -1;
    }

    slconn->begin_time = strdup (timelist->element);

    timelist = timelist->next;

    if (timelist != 0)
    {
      slconn->end_time = strdup (timelist->element);

      if (timelist->next != 0)
      {
        sl_log (2, 0, "malformed time window specification\n");
        sl_strparse (NULL, NULL, &timelist);
        return -1;
      }
    }

    /* Free the parsed list */
    sl_strparse (NULL, NULL, &timelist);
  }

  /* Parse the 'multiselect' string following '-S' */
  if (multiselect)
  {
    if (sl_parse_streamlist (slconn, multiselect, selectors) == -1)
      return -1;
  }
  else if (!streamfile)
  { /* No 'streams' array, assuming uni-station mode */
    sl_setuniparams (slconn, selectors, -1, 0);
  }

  /* Attempt to recover sequence numbers from state file */
  if (statefile)
  {
    /* Check if interval was specified for state saving */
    if ((tptr = strchr (statefile, ':')) != NULL)
    {
      char *tail;

      *tptr++ = '\0';

      stateint = (unsigned int)strtoul (tptr, &tail, 0);

      if (*tail || (stateint < 0 || stateint > 1e9))
      {
        sl_log (2, 0, "state saving interval specified incorrectly\n");
        return -1;
      }
    }

    if (sl_recoverstate (slconn, statefile) < 0)
    {
      sl_log (2, 0, "state recovery failed\n");
    }
  }

  return 0;
} /* End of parameter_proc() */

/***************************************************************************
 * getoptval:
 *
 * Return the value to a command line option; checking that the value is 
 * itself not an option (starting with '-') and is not past the end of
 * the argument list.
 *
 * argcount: total arguments in argvec
 * argvec: argument list
 * argopt: index of option to process, value is expected to be at argopt+1
 *
 * Returns value on success and exits with error message on failure
 ***************************************************************************/
static char *
getoptval (int argcount, char **argvec, int argopt)
{
  if (argvec == NULL || argvec[argopt] == NULL)
  {
    fprintf (stderr, "getoptval(): NULL option requested\n");
    exit (1);
  }

  if ((argopt + 1) < argcount && *argvec[argopt + 1] != '-')
    return argvec[argopt + 1];

  fprintf (stderr, "Option %s requires a value\n", argvec[argopt]);
  exit (1);
} /* End of getoptval() */

/***************************************************************************
 * term_handler:
 * Signal handler routine to control termination.
 ***************************************************************************/
static void
term_handler (int sig)
{
  sl_terminate (slconn);
}

/***************************************************************************
 * print_timelogc:
 *
 * Log message print handler used with sl_loginit() and sl_log().  Prefixes
 * a local time string to the message before printing.
 ***************************************************************************/
static void
print_timelogc (const char *msg)
{
  char timestr[100];
  time_t loc_time;

  /* Build local time string and cut off the newline */
  time (&loc_time);
  strcpy (timestr, asctime (localtime (&loc_time)));
  timestr[strlen (timestr) - 1] = '\0';

  fprintf (stdout, "%s - %s", timestr, msg);
}

/***************************************************************************
 * print_timelog:
 *
 * A wrapper for print_timelog() with a non-const msg.
 ***************************************************************************/
static void
print_timelog (char *msg)
{
  print_timelogc (msg);
}

/***************************************************************************
 * usage:
 * Print the usage message and exit.
 ***************************************************************************/
static void
usage (void)
{
  fprintf (stderr, "%s version %s\n\n", PACKAGE, VERSION);
  fprintf (stderr, "Usage: %s [options] slhost dlhost\n\n", PACKAGE);
  fprintf (stderr,
           " ## General options ##\n"
           " -V              Report program version\n"
           " -h              Print this usage message\n"
           " -v              Be more verbose, multiple flags can be used\n"
           " -d              Configure SeedLink connection in dial-up mode\n"
           " -nd delay       network re-connect delay (seconds), default 30\n"
           " -nt timeout     network timeout (seconds), re-establish connection if no\n"
           "                   data/keepalives are received in this time, default 600\n"
           " -k interval     send keepalive (heartbeat) packets this often (seconds)\n"
           " -N netcode      Change all SEED network codes to specified code\n"
           " -x sfile[:int]  Save/restore stream state information to this file\n"
           "\n"
           " ## SeedLink data stream selection ##\n"
           " -s selectors    Selectors for uni/all-station or default for multi-station mode\n"
           " -l listfile     Read a stream list from this file for multi-station mode\n"
           " -S streams      Define a stream list for multi-station mode\n"
           "   'streams' = 'stream1[:selectors1],stream2[:selectors2],...'\n"
           "        'stream' is in NET_STA format, for example:\n"
           "        -S \"IU_KONO:BHE BHN,GE_WLF,MN_AQU:HH?.D\"\n\n"
           " -tw begin:[end]  (requires SeedLink >= 3)\n"
           "        specify a time window in year,month,day,hour,min,sec format\n"
           "        example: -tw 2002,08,05,14,00,00:2002,08,05,14,15,00\n"
           "        the end time is optional, but the colon must be present\n"
           "\n"
           " slhost   Address of the SeedLink server in host:port format\n"
           "            Default host is 'localhost' and default port is '18000'\n\n"
           " dlhost   Address of the DataLink server in host:port format\n"
           "            Default host is 'localhost' and default port is '16000'\n\n");

} /* End of usage() */
